src/scaladoc/scala/tools/nsc/ScalaDoc.scala

```diff
@@ -38,23 +39,50 @@ class ScalaDoc {
       reporter.echo(command.usageMsg)
     else
       try { new DocFactory(reporter, docSettings) document command.files }
-    catch {
-      case ex @ FatalError(msg) =>
-        if (docSettings.debug.value) ex.printStackTrace()
-        reporter.error(null, "fatal error: " + msg)
-    }
-    finally reporter.printSummary()
+      catch {
+        case ex @ FatalError(msg) =>
+          if (docSettings.debug.value) ex.printStackTrace()
+          reporter.error(null, "fatal error: " + msg)
+      }
+      finally reporter.printSummary()
 
     !reporter.reallyHasErrors
   }
 }
 
+/** The Scaladoc reporter adds summary messages to the `ConsoleReporter`
+ *
+ *  Use the `summaryX` methods to add unique summarizing message to the end of
+ *  the run.
+ */
 class ScalaDocReporter(settings: Settings) extends ConsoleReporter(settings) {
+  import scala.collection.mutable
 
   // need to do sometimes lie so that the Global instance doesn't
   // trash all the symbols just because there was an error
   override def hasErrors = false
   def reallyHasErrors = super.hasErrors
+
+  private[this] implicit val orderingByPos: Ordering[Position] = Ordering.by(p => if (p.isDefined) p.start else -1)
+  private[this] val delayedMessages: mutable.SortedMap[Position, mutable.Map[String, () => Unit]] =
+    mutable.SortedMap.empty[Position, mutable.Map[String, () => Unit]]
+
+  /** Eliminates messages if both `pos` and `msg` are equal to existing element */
+  def addDelayedMessage(pos: Position, msg: String, print: () => Unit): Unit =
+    delayedMessages.get(pos).map { posMap =>
+      posMap += (msg -> print)
+    } getOrElse {
+      delayedMessages(pos) = mutable.Map(msg -> print)
+    }
+
+  def printDelayedMessages(): Unit = for {
+    (_, map) <- delayedMessages
+  } yield map.values.foreach(_.apply())
```

Comment:
actually you could even do a `LinkedHashMap[(Position, String), () => Unit]`, as the insertion order should be deterministic, wdyt?